File: hooks/registry.py
"""Lightweight event registry for the Hook system."""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any, Callable


@dataclass
class HookEvent:
    """Hook event carrier."""

    name: str
    data: dict[str, Any] = field(default_factory=dict)


class HookRegistry:
    """Lightweight event registration and dispatch center."""

    def __init__(self) -> None:
        self._hooks: dict[str, list[Callable[[HookEvent], None]]] = {}
        self._log: Any = None

    def bind_log(self, log: Any) -> None:
        """Bind a FlowLog instance. All events auto-forward to log.step()."""
        self._log = log

    def register(self, event: str, hook: Callable[[HookEvent], None]) -> None:
        """Register a custom hook for an event."""
        self._hooks.setdefault(event, []).append(hook)

    def fire(self, event: str, **data: Any) -> None:
        """
        Fire an event (synchronous).
        1. Forward to FlowLog.step() if bound.
        2. Call all registered custom hooks.
        """
        if self._log is not None:
            self._log.step(event, **data)

        hook_event = HookEvent(name=event, data=data)
        for hook in self._hooks.get(event, []):
            try:
                hook(hook_event)
            except Exception:
                pass  # one failing hook does not affect others

File: src/hooks/registry.ts
/**
 * HookRegistry — 轻量级事件注册和分发中心
 * 实现 orchestrator 层的 HookRegistry 接口
 */

import type { FlowLog } from '../log/flow.js'
import type { HookEvent } from '../types.js'

export class HookRegistry {
  private log: FlowLog | null = null
  private hooks: Map<string, Array<(event: HookEvent) => void | Promise<void>>> = new Map()

  /** 绑定 FlowLog 实例，所有事件自动转发到 log.step() */
  bindLog(log: FlowLog): void {
    this.log = log
  }

  /** 注册自定义 hook */
  register(event: string, hook: (event: HookEvent) => void | Promise<void>): void {
    const existing = this.hooks.get(event) ?? []
    existing.push(hook)
    this.hooks.set(event, existing)
  }

  /** 触发事件：先转发到 FlowLog，再调用所有注册的 hook（单个异常不影响其他） */
  async fire(name: string, data: Record<string, unknown> = {}): Promise<void> {
    // 1. 转发到 FlowLog
    if (this.log) {
      try {
        this.log.step(name, data)
      } catch (e) {
        console.warn('HookRegistry: log.step() failed:', e instanceof Error ? e.message : e)
      }
    }

    // 2. 调用注册的 hook
    const hookEvent: HookEvent = { name, data }
    const hooks = this.hooks.get(name) ?? []
    for (const hook of hooks) {
      try {
        await hook(hookEvent)
      } catch (e) {
        console.warn(`HookRegistry: hook failed for '${name}':`, e instanceof Error ? e.message : e)
      }
    }
  }
}
